### ISSUE TITLE
Cannot switch from httpapi connection plugin to network_cli in the same workflow

### ISSUE THREAD

**Issue by jean-christophe-manciot** [ansible-collections/ansible.netcommon#45]
##### SUMMARY
When an attempt is made to use ```ansible_connection: network_cli```  after having used ```ansible_connection: httpapi``` connection plugin with the same NXOS device in the same workflow, the latter is used instead.

This is a fatal issue with some networking modules - [such as nxos_file_copy or net_put](https://github.com/ansible-collections/cisco.nxos/issues/75#issue-627411476) - which cannot be used with such httpapi connection plugin.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
httpapi

##### ANSIBLE VERSION
```
ansible 2.9.9
  config file = /etc/ansible/ansible.cfg
  ansible python module location = /usr/local/lib/python3.8/dist-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.8.3 (default, May 14 2020, 22:08:57) [GCC 9.3.0]
```

##### CONFIGURATION
```
ANSIBLE_PIPELINING(/etc/ansible/ansible.cfg) = True
CACHE_PLUGIN(/etc/ansible/ansible.cfg) = redis
CACHE_PLUGIN_TIMEOUT(/etc/ansible/ansible.cfg) = 3600
DEFAULT_EXECUTABLE(/etc/ansible/ansible.cfg) = /bin/bash
DEFAULT_FORKS(/etc/ansible/ansible.cfg) = 1000
DEFAULT_GATHERING(/etc/ansible/ansible.cfg) = explicit
DEFAULT_GATHER_TIMEOUT(/etc/ansible/ansible.cfg) = 30
DEFAULT_HASH_BEHAVIOUR(/etc/ansible/ansible.cfg) = merge
DEFAULT_HOST_LIST(/etc/ansible/ansible.cfg) = ['/etc/ansible/hosts']
DEFAULT_LOG_PATH(/etc/ansible/ansible.cfg) = /var/log/ansible.log
DEFAULT_PRIVATE_ROLE_VARS(/etc/ansible/ansible.cfg) = False
DEFAULT_TIMEOUT(/etc/ansible/ansible.cfg) = 180
DEFAULT_TRANSPORT(/etc/ansible/ansible.cfg) = ssh
ENABLE_TASK_DEBUGGER(/etc/ansible/ansible.cfg) = True
HOST_KEY_CHECKING(/etc/ansible/ansible.cfg) = True
INTERPRETER_PYTHON(/etc/ansible/ansible.cfg) = /usr/bin/python3
PERSISTENT_COMMAND_TIMEOUT(/etc/ansible/ansible.cfg) = 3599
PERSISTENT_CONNECT_RETRY_TIMEOUT(/etc/ansible/ansible.cfg) = 200
PERSISTENT_CONNECT_TIMEOUT(/etc/ansible/ansible.cfg) = 3600
RETRY_FILES_ENABLED(/etc/ansible/ansible.cfg) = False
SHOW_CUSTOM_STATS(/etc/ansible/ansible.cfg) = True
```

##### OS / ENVIRONMENT
```
controller host: Ubuntu focal 20.04
paramiko : 2.7.1

remote host: NX_OSv_9k 9.2.3
```

##### STEPS TO REPRODUCE
```
- name: Setting hostname on 9k
  vars:
        ansible_connection: httpapi
        ansible_httpapi_password: 'password'
        ansible_httpapi_port: 443
        ansible_httpapi_remote_user: admin
        ansible_httpapi_use_ssl: yes
        ansible_httpapi_validate_certs: no
        ansible_network_os: nxos
  nxos_config:
      replace: line
      save_when: modified
      lines:
        - "hostname {{ inventory_hostname }}"

- name: Pushing some file
  vars:
        ansible_become: yes
        ansible_become_method: enable
        ansible_become_password: 'password'
        ansible_connection: network_cli
        ansible_network_os: nxos
        ansible_password: 'password'
        ansible_user: admin
  nxos_file_copy:
        local_file: /some/file
```

##### EXPECTED RESULTS
Successful:
- hostname change with ```ansible_connection: httpapi```
- file push with ```ansible_connection: network_cli```

##### ACTUAL RESULTS
- Successful hostname change with ```ansible_connection: httpapi```
- Failed file push with ```ansible_connection: network_cli```

```
Using module file /usr/local/lib/python3.8/dist-packages/ansible/modules/network/nxos/nxos_config.py
Pipelining is enabled.
<172.21.97.178> ESTABLISH LOCAL CONNECTION FOR USER: admin
<172.21.97.178> EXEC /bin/bash -c '/usr/bin/python3 && sleep 0'
ok: [NX_OSv_9k-9.2.3] => {
    "changed": false,
    "invocation": {
        "module_args": {
            "after": null,
            "auth_pass": null,
            "authorize": null,
            "backup": false,
            "backup_options": null,
            "before": null,
            "defaults": false,
            "diff_against": null,
            "diff_ignore_lines": null,
            "host": null,
            "intended_config": null,
            "lines": [
                "hostname NX_OSv_9k-9.2.3"
            ],
            "match": "line",
            "parents": null,
            "password": null,
            "port": null,
            "provider": null,
            "replace": "line",
            "replace_src": null,
            "running_config": null,
            "save_when": "modified",
            "src": null,
            "ssh_keyfile": null,
            "timeout": null,
            "transport": null,
            "use_ssl": null,
            "username": null,
            "validate_certs": null
        }
    }
}

<172.21.97.178> attempting to start connection
<172.21.97.178> using connection plugin httpapi
<172.21.97.178> found existing local domain socket, using it!
<172.21.97.178> updating play_context for connection
<172.21.97.178> 
<172.21.97.178> local domain socket path is home/user/.ansible/pc/daf7601ce6
fatal: [NX_OSv_9k-9.2.3]: FAILED! => {
    "changed": false,
    "msg": "Connection type must be fully qualified name for network_cli connection type, got httpapi"
}
```

**Comment by Qalthos** [comment]
So first, I don't know if you're setting vars on the tasks just to be demonstrative or if you're really doing this all the time. If it's the former, I really *genuinely* appreciate that, as this makes it a whole lot easier to know exactly what's going on. But just in case it's the latter (or someone else sees this and thinks this is a neat way to specify connection properties), you _really_ should use inventory to keep all those straight. The biggest reason we moved to `network_cli`/`httpapi`/`netconf` connections was so you didn't have to specify them every time.

Alright, so with that out of the way, let's dig in to this. I made a standalone playbook out of your snippet:
```
---
- hosts: nxos_no_vars
  gather_facts: no  # no connection set, so we can't get facts

  tasks:
    - name: Setting hostname on 9k
      vars:
        ansible_connection: httpapi
        ansible_httpapi_password: 'password'
        ansible_httpapi_port: 443
        ansible_user: admin
        ansible_httpapi_use_ssl: yes
        ansible_httpapi_validate_certs: no
        ansible_network_os: nxos
      nxos_config:
        replace: line
        save_when: modified
        lines:
          - "hostname {{ inventory_hostname }}"

    - name: Pushing some file
      vars:
        ansible_become: yes
        ansible_become_method: enable
        ansible_become_password: 'password'
        ansible_connection: network_cli
        ansible_network_os: nxos
        ansible_password: 'password'
        ansible_user: admin
      nxos_file_copy:
        local_file: empty.file
```

And it gave me `401 Authorization Required` on the first task. This is because `ansible_httpapi_remote_user` is _not_ an option that httpapi recognizes... the only thing it takes is plain old `ansible_user` (along with some internal magic that reroutes other names to that var that it is not worth going into here). With that in mind, I deduplicated some of the vars as follows:

```
---
- hosts: nxos_no_vars
  gather_facts: no  # no connection set, so we can't get facts

  vars:
    ansible_user: admin
    ansible_password: 'password'
    ansible_network_os: nxos

  tasks:
    - name: Setting hostname on 9k
      vars:
        ansible_connection: httpapi
        ansible_httpapi_port: 443
        ansible_httpapi_use_ssl: yes
        ansible_httpapi_validate_certs: no
      nxos_config:
        replace: line
        save_when: modified
        lines:
          - "hostname {{ inventory_hostname }}"

    - name: Pushing some file
      vars:
        ansible_connection: network_cli
        ansible_become: yes
        ansible_become_method: enable
        ansible_become_password: 'password'
      nxos_file_copy:
        local_file: empty.file
```

and got this in response:

```
PLAY [nxos_no_vars] ***************************************************************************************************

TASK [Setting hostname on 9k] *****************************************************************************************
ok: [nxosv]

TASK [Pushing some file] **********************************************************************************************
changed: [nxosv]

PLAY RECAP ************************************************************************************************************
nxosv                      : ok=2    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

So if you are in fact using that playbook as written, then you must have at least some variables set outside of the tasks. a quick look at [variable precedence](https://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html#variable-precedence-where-should-i-put-a-variable) shows a very short list of things that beat task vars. To wit:

* task vars (only for the task)
* include_vars
* set_facts / registered vars
* role (and include_role) params
* include params
* extra vars (always win precedence)

So, upon adding a `set_fact` to the playbook:

```
---
- hosts: nxos
  gather_facts: no  # no connection set, so we can't get facts

  vars:
    ansible_user: admin
    ansible_password: 'password'
    ansible_network_os: nxos

  tasks:
    - set_fact:
        ansible_connection: httpapi

    - name: Setting hostname on 9k
      vars:
        ansible_connection: httpapi
        ansible_httpapi_port: 443
        ansible_httpapi_use_ssl: yes
        ansible_httpapi_validate_certs: no
      nxos_config:
        replace: line
        save_when: modified
        lines:
          - "hostname {{ inventory_hostname }}"

    - name: Pushing some file
      vars:
        ansible_connection: network_cli
        ansible_become: yes
        ansible_become_method: enable
        ansible_become_password: 'password'
      nxos_file_copy:
        local_file: empty.file
```

I now see the reported issue:

```
PLAY [nxos] ***********************************************************************************************************

TASK [set_fact] *******************************************************************************************************
ok: [nxosv]

TASK [Setting hostname on 9k] *****************************************************************************************
ok: [nxosv]

TASK [Pushing some file] **********************************************************************************************
fatal: [nxosv]: FAILED! => {"changed": false, "msg": "Connection type must be fully qualified name for network_cli connection type, got httpapi"}

PLAY RECAP ************************************************************************************************************
nxosv                      : ok=2    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   
```

So... check to see if you've set `ansible_connection` elsewhere. I couldn't speculate on which might be the issue, but the good news is that the list is pretty short and it shouldn't take too long to track down.

**Comment by jean-christophe-manciot** [comment]
> So first, I don't know if you're setting vars on the tasks just to be demonstrative or if you're really doing this all the time.

I use inventory group_vars & host_vars extensively; setting vars here was to make all readers'  life easier.

> This is because `ansible_httpapi_remote_user` is _not_ an option that httpapi recognizes... the only thing it takes is plain old `ansible_user`

So there is ```ansible_httpapi_password``` but not ```ansible_httpapi_remote_user```...
Anyway, it's not really an issue because ```ansible_user``` has already been set in my hosts file.

> So... check to see if you've set `ansible_connection` elsewhere. I couldn't speculate on which might be the issue, but the good news is that the list is pretty short and it shouldn't take too long to track down.

There is an inventory group_vars file which sets all connection parameters.
According to the [variables precedence](https://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html#variable-precedence-where-should-i-put-a-variable), it cannot be a problem:
```
...
4. inventory group_vars/all
...
17. task vars (only for the task)
```
You were able to replicate the issue, so there is nothing more special with my use case which could trigger that issue.

**Comment by jean-christophe-manciot** [comment]
This issue may be related to some variables precedence issue.
It is said in the official doc that:
```
15. role vars (defined in role/vars/main.yml)
...
20. role (and include_role) params
```
My use case is a little bit more complex that what my first post may suggest.
Since the variables are actually set (with ```vars:``` as displayed in the first post) within different roles:
- the priority 15 should not apply
- the priority 20 should apply

If we consider a continuous workflow calling multiple roles, it should be possible to assign the same variable (```ansible_connection``` for instance) to 2 different values  between 2 different roles or within the same role, right?

I have a feeling that the issue may originate from a variable which cannot be reassigned at the same "priority"  (20 here).
I hope this is not an expected behavior.

**Comment by jean-christophe-manciot** [comment]
Also, I never use ```set_fact``` to modify those kind of ```ansible_``` connections variables within roles.

**Comment by jean-christophe-manciot** [comment]
To be more precise, the last failed task is called within an included "sub-role":
**role/tasks/main.ym**:
```
- include: "sub-role.yml"
```
**role/tasks/sub-role.ym**:
```
- name: Pushing some file
  vars:
        ansible_become: yes
        ansible_become_method: enable
        ansible_become_password: 'password'
        ansible_connection: network_cli
        ansible_network_os: nxos
        ansible_password: 'password'
        ansible_user: admin
  nxos_file_copy:
        local_file: /some/file
```
Is it fair to assume that the priority of those vars are also **17**?

**Comment by Qalthos** [comment]
I will preface this by saying that I am not an expert in variable precedence, and I am taking the variable precedence list as read. I also can't see your playbooks, so I cannot say for sure what is going on.

>Is it fair to assume that the priority of those vars are also 17?

Those should still be task vars, yes. Their being contained within a role does not expand their scope somehow.

>I have a feeling that the issue may originate from a variable which cannot be reassigned at the same "priority" (17 here).

Again, I don't have your playbooks, so I can't say with any specificity, but if there is confusion about what value should apply where, then you are probably setting the same variable too often. vars have the same scope as the construct to which they are attached, and (with the exception detailed below) are overridden in order of specificity. For `connection`, there should very little reason to have any confusion over which variable is in force: either the one in inventory, or the one attached to the task.

I did mention an exception to this, and my guess is that include params are probably where the issue is, because they can be confusingly broad in scope. https://github.com/ansible-collections/arista.eos/blob/master/tests/integration/targets/eos_interfaces/tasks/eapi.yaml#L26 is what it looks like when we use include params in our testing to run the tests on each connection type. Specifically,

```
include: '{{ test_case_to_run }} ansible_connection=ansible.netcommon.httpapi'
```

though as that syntax is deprecated they should probably now read

```
include: '{{ test_case_to_run }}'
vars:
  ansible_connection: ansible.netcommon.httpapi
```

Note how this looks just like a task var! In either case, every task in `test_case_to_run` will have the `httpapi` connection set on it, and due to the way `include` works the _only_ thing that can override that is `-e` on the command line, which is the complete opposite of what you want.

All that said, I don't know how else to help you. If you cannot produce a minimal set of playbooks that reproduces what you're seeing, there really isn't anything I can do other than speculate. Nothing I have seen in my testing indicates that variable precedence is working incorrectly- and I actually learned a lot about how it works trying to figure out what could possibly be happening. But without something to go on, I don't see how else I can help you other than "you're setting the variable somewhere else and you don't realize it".

**Comment by jean-christophe-manciot** [comment]
My own suspicion regarding the role of```include```  in this issue was confirmed when I replaced it with ```include_tasks```, which ... worked around the issue. 

This ```include``` seems to have some other strange/unintuitive behaviors as mentioned [here](https://docs.ansible.com/ansible/latest/modules/include_module.html#notes).
Since it is to be deprecated sometime in the future, it seems unreasonable to try to solve this issue.
Anyway, I have upgraded all my ancient ```include``` to the more appropriate ```include_tasks``` in all my roles.
